### PR TITLE
Implement a proper generic resolution in `decl_storage!`

### PR DESCRIPTION
**Pull request by bkchr**
Until now, we are required to add the generic type to types generated by `decl_storage!` that do not require it.
The worst outcome of this is the requirement of `_genesis_phantom_data` in `GenesisConfig`.

All of this is now fixed by this pr. We do not require anymore a phantom field in a `GenesisConfig`, as we do a proper lookup if a generic type is required.

Fixes: #2219

### CODE REVIEW

**Comment by thiolliere** [comment]
as a note the API change in the following way:
* storage structs that doesn't need generics are now not generic: `<CandidateCount<Test>>::put(1);` is now `CandidateCount::put(1);` for some storages this breaks homogeneity of storages but make more sense with the underlying implementation.
* `GenesisConfig` structs that doesn't need generics are now not generic as well, same as above this break homogeneity in favor of actual implementation
* `GenesisConfig` no longer implements BuildStorage trait but only some methods which signatures differs depending on if GenesisConfig is generic of not:
  `system::GenesisConfig::<Test>::default().build_storage().unwrap().0.into()` is now `system::GenesisConfig::default().build_storage::<Test>().unwrap().0.into()` for some config.

**Comment by bkchr** [comment]
Thx for the extended summary :D

> as a note the API change in the following way:
> 
>     * storage structs that doesn't need generics are now not generic: `<CandidateCount<Test>>::put(1);` is now `CandidateCount::put(1);` for some storages this breaks homogeneity of storages but make more sense with the underlying implementation.
> 

Yeah, I wanted to have it the same way, as it would be in a manual implementation.

>     * `GenesisConfig` structs that doesn't need generics are now not generic as well, same as above this break homogeneity in favor of actual implementation
> 

This does not break homogeneity, we already could have `GenesisConfig`s without a generic argument. Now, we just do it the right way. Before, if no generic was required, but a build function was given, we still required the generic.
 
>     * `GenesisConfig` no longer implements BuildStorage trait but only some methods which signatures differs depending on if GenesisConfig is generic of not:
>       `system::GenesisConfig::<Test>::default().build_storage().unwrap().0.into()` is now `system::GenesisConfig::default().build_storage::<Test>().unwrap().0.into()` for some config.

You could also have `system::GenesisConfig::default().build_storage::<Test, Instance>().unwrap().0.into()`. This is sadly required, as we can have build function that use some generics.

File: srml/support/test/tests/issue2219.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			/// Bonding only occurs after successful entry into a role.
+			pub RoleEntryRequests get(role_entry_requests) : Requests<T>;
+
+			/// Entry request expires after this number of blocks
+			pub RequestLifeTime get(request_life_time) config(request_life_time) : u64 = 0;
+		}
+		add_extra_genesis {
+			config(enable_storage_role): bool;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (bkchr) [reply srml/support/test/tests/issue2219.rs]: Could be easier, but I don't see any problem of having all of it. It is just a test that makes sure that nothing breaks ;)

File: srml/support/test/tests/instance.rs [hunk 2]
```diff
@@ -326,48 +270,48 @@ fn new_test_ext() -> runtime_io::TestExternalities<Blake2Hasher> {
...
-			module2::LinkedMap::<Runtime, module2::Instance3>::key_for(1).to_vec(),
-			module2::DoubleMap::<Runtime>::prefix_for(1),
-			module2::DoubleMap::<Runtime, module2::Instance1>::prefix_for(1).to_vec(),
-			module2::DoubleMap::<Runtime, module2::Instance2>::prefix_for(1).to_vec(),
-			module2::DoubleMap::<Runtime, module2::Instance3>::prefix_for(1).to_vec(),
-			module2::DoubleMap::<Runtime>::key_for(1, 1),
-			module2::DoubleMap::<Runtime, module2::Instance1>::key_for(1, 1).to_vec(),
-			module2::DoubleMap::<Runtime, module2::Instance2>::key_for(1, 1).to_vec(),
-			module2::DoubleMap::<Runtime, module2::Instance3>::key_for(1, 1).to_vec(),
+			module2::Map::<module2::DefaultInstance>::prefix().to_vec(),
```
[the remark this answers is not in the document]
Comment (bkchr) [reply srml/support/test/tests/instance.rs]: The compiler wants it now from you, but that should be no problem. In a real module, you always have to give the Instance generic anyway. So, this is no problem.

File: srml/support/src/runtime.rs [hunk 3]
```diff
@@ -140,10 +141,12 @@ macro_rules! construct_runtime {
 			$( $rest )*
 		);
 	};
+	// `default` identifier given, expand to default + given extra modules
 	(
 		{ $( $preset:tt )* };
 		{ $( $expanded:tt )* };
 		$name:ident: $module:ident::{
+			default,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply srml/support/src/runtime.rs]: Good find, the above one is incorrect and is probably a merge bug.

File: srml/support/procedural/src/storage/transformation.rs [hunk 4]
```diff
@@ -234,9 +245,10 @@ fn decl_store_extra_genesis(
 					)
 				);
 			};
-			if type_infos.kind.is_simple() && ext::has_parametric_type(type_infos.value_type, traitinstance) {
+
+			if type_infos.kind.is_simple()
+				&& ext::type_contains_ident(type_infos.value_type, traitinstance) {
```
Comment (thiolliere) [inline srml/support/procedural/src/storage/transformation.rs]: this was already wrong but we have to set is_trait_needed for other kind as well (see test attached)

[t.txt](https://github.com/paritytech/substrate/files/3313955/t.txt)

File: srml/support/procedural/src/storage/transformation.rs [hunk 5]
```diff
@@ -281,48 +291,79 @@ fn decl_store_extra_genesis(
 				}).unwrap_or_else(|| quote!( Default::default() ));
 
 			config_field_default.extend(quote!( #ident: #fielddefault, ));
+
+			opt_build
+				.map(Some)
+				.unwrap_or_else(||
```
Comment (thiolliere) [inline srml/support/procedural/src/storage/transformation.rs]: use `or_else`

File: srml/support/procedural/src/storage/transformation.rs [hunk 6]
```diff
@@ -281,48 +291,79 @@ fn decl_store_extra_genesis(
...
-			opt_build = build.inner.as_ref().map(|b| &b.expr.content).map(|b| quote!( #b ));
-		}
+			opt_build
+		};
 
 		let typ = type_infos.typ;
-		if let Some(builder) = opt_build {
-			is_trait_needed = true;
+		if let Some(builder) = builder {
+			assimilate_require_generic = true;
```
Comment (thiolliere) [inline srml/support/procedural/src/storage/transformation.rs]: is that an error ? then you reset it to true again for some value in the match

Comment (bkchr) [reply srml/support/procedural/src/storage/transformation.rs]: Hmm yeah.

File: srml/support/src/origin.rs [hunk 7]
```diff
@@ -72,33 +93,80 @@ macro_rules! impl_outer_origin {
...
+			$(#[$attr])*
+			#[allow(non_camel_case_types)]
+			pub enum $name {
+				system($system::Origin<$runtime>),
+				$(
+					[< $module $( _ $generic_instance )? >]
+					($module::Origin < $( $generic, )? $( $module::$generic_instance )? > ),
+				)*
+				#[allow(dead_code)]
+				Void($crate::Void)
```
[the remark this answers is not in the document]
Comment (thiolliere) [reply srml/support/src/origin.rs]: I mean I removed it and all test pass

[the remark this answers is not in the document]
Comment (bkchr) [reply srml/support/src/origin.rs]: Good question, but I don't really know it and would rather like to keep it for the moment.^^

**Comment by bkchr** [comment]
@thiolliere addressed your feedback :)

**Comment by thiolliere** [comment]
Yes changes looks good I can approve or do a final review on tuesday.

Also we might want to update decl_storage doc, as the stucture is no longer always generic.

And I was thinking does it make sense to talk about assimilate storage stuff ? I mean it is for used when writing test for your modules, but I think it is something that ppl want to know, though I'm not sure and could added in a second time as it doesn't exist yet.

**Comment by bkchr** [comment]
> Yes changes looks good I can approve or do a final review on tuesday.
> 

Tuesday is okay. 

> Also we might want to update decl_storage doc, as the stucture is no longer always generic.
> 

Good point. 
> And I was thinking does it make sense to talk about assimilate storage stuff ? I mean it is for used when writing test for your modules, but I think it is something that ppl want to know, though I'm not sure and could added in a second time as it doesn't exist yet.

You mean the function `assimilate_storage`?

**Comment by gavofyork** [comment]
did the final review happen?

**Comment by bkchr** [comment]
Not yet.

**Comment by thiolliere** [comment]
>>  And I was thinking does it make sense to talk about assimilate storage stuff ? I mean it is for used when writing test for your modules, but I think it is something that ppl want to know, though I'm not sure and could added in a second time as it doesn't exist yet.
>
> You mean the function assimilate_storage?

Yes I know construct_runtime handle that you, but someone that create a modules will make use of it for its testing so he wants to be aware of subtelties (though everything is in the generated doc)

**Pull request merged**